### ISSUE TITLE
cannot bootstrap models receiving NoValidClassError

### ISSUE THREAD

**Issue by nhhockeyplayer** [typegoose/typegoose#638]
I bootstrap my nest api server which launches fine 
if I do not import my model module launches fine

if I import my model module, I receive the following console error

```
> nx run api-typegoose:serve 
chunk (runtime: main) main.js (main) 29.1 KiB [entry] [rendered]
webpack compiled successfully (6acadf051197bc21)
Debugger listening on ws://localhost:9229/ef6541ae-32df-4b3f-83b1-88137a076712
Debugger listening on ws://localhost:9229/ef6541ae-32df-4b3f-83b1-88137a076712
For help, see: https://nodejs.org/en/docs/inspector
Issues checking in progress...

/Users/meanstack02/new/starter/dist/apps/api-typegoose/webpack:/starter/libs/shared-entity/src/lib/typegoose/abstract.entity.ts:60
        return this._id.toString()
                        ^
TypeError: Cannot read properties of undefined (reading 'toString')
    at Integrity.hash (/Users/meanstack02/new/starter/dist/apps/api-typegoose/webpack:/starter/libs/shared-entity/src/lib/typegoose/abstract.entity.ts:60:25)
    at Integrity.toString (/Users/meanstack02/new/starter/dist/apps/api-typegoose/webpack:/starter/libs/shared-entity/src/lib/typegoose/abstract.entity.ts:64:21)
    at new NoValidClassError (/Users/meanstack02/new/starter/node_modules/@typegoose/typegoose/src/internal/errors.ts:36:15)
    at Object.getName (/Users/meanstack02/new/starter/node_modules/@typegoose/typegoose/src/internal/utils.ts:313:37)
    at processProp (/Users/meanstack02/new/starter/node_modules/@typegoose/typegoose/src/internal/processProp.ts:33:22)
    at _buildSchema (/Users/meanstack02/new/starter/node_modules/@typegoose/typegoose/src/internal/schema.ts:59:18)
    at buildSchema (/Users/meanstack02/new/starter/node_modules/@typegoose/typegoose/src/typegoose.ts:148:23)
    at Object.getModelForClass (/Users/meanstack02/new/starter/node_modules/@typegoose/typegoose/src/typegoose.ts:78:58)
    at Object../libs/shared-entity/src/lib/typegoose/role.entity.ts (/Users/meanstack02/new/starter/dist/apps/api-typegoose/webpack:/starter/libs/shared-entity/src/lib/typegoose/role.entity.ts:34:19)
    at __webpack_require__ (/Users/meanstack02/new/starter/dist/apps/api-typegoose/webpack:/starter/webpack/bootstrap:19:1)
    at Object../apps/api-typegoose/src/app/domain-model/role/role.controller.ts (/Users/meanstack02/new/starter/dist/apps/api-typegoose/webpack:/starter/apps/api-typegoose/src/app/domain-model/role/role.controller.ts:6:1)
    at __webpack_require__ (/Users/meanstack02/new/starter/dist/apps/api-typegoose/webpack:/starter/webpack/bootstrap:19:1)
    at Object../apps/api-typegoose/src/app/domain-model/role/role.module.ts (/Users/meanstack02/new/starter/dist/apps/api-typegoose/webpack:/starter/apps/api-typegoose/src/app/domain-model/role/role.module.ts:2:1)
    at __webpack_require__ (/Users/meanstack02/new/starter/dist/apps/api-typegoose/webpack:/starter/webpack/bootstrap:19:1)
    at Object../apps/api-typegoose/src/app/app.module.ts (/Users/meanstack02/new/starter/dist/apps/api-typegoose/webpack:/starter/apps/api-typegoose/src/app/app.module.ts:7:1)
    at __webpack_require__ (/Users/meanstack02/new/starter/dist/apps/api-typegoose/webpack:/starter/webpack/bootstrap:19:1)
No issues found.



```
## Versions

` "@typegoose/typegoose": "^9.2.0",`

code is as follows which uses nestjs-typegoose

```
@Module({
    imports: [
        ConfigModule.forRoot({
            cache: true
        }),

        TypegooseModule.forRoot(dbConnectString, {
        }),
        RoleModule,

```

```
meanstack02@admin starter % nx report

>  NX  Report complete - copy this into the issue template

  Node : 16.13.0
  OS   : darwin x64
  yarn : 1.22.17
  
  nx : 13.1.4
  @nrwl/angular : 13.1.4
  @nrwl/cli : 13.1.4
  @nrwl/cypress : 13.1.4
  @nrwl/devkit : 13.1.4
  @nrwl/eslint-plugin-nx : 13.1.4
  @nrwl/express : 13.1.4
  @nrwl/jest : 13.1.4
  @nrwl/linter : 13.1.4
  @nrwl/nest : 13.1.4
  @nrwl/next : undefined
  @nrwl/node : 13.1.4
  @nrwl/nx-cloud : 12.5.2
  @nrwl/react : 13.1.4
  @nrwl/react-native : undefined
  @nrwl/schematics : 8.12.11
  @nrwl/tao : 13.1.4
  @nrwl/web : 13.1.4
  @nrwl/workspace : 13.2.3
  @nrwl/storybook : 13.1.4
  @nrwl/gatsby : undefined
  typescript : 4.3.5
  rxjs : 7.4.0
  ---------------------------------------
  Community plugins:
  	 @angular/animations: 12.2.13
  	 @angular/cdk: 12.2.13
  	 @angular/common: 12.2.13
  	 @angular/compiler: 12.2.13
  	 @angular/core: 12.2.13
  	 @angular/forms: 12.2.13
  	 @angular/material: 12.2.13
  	 @angular/platform-browser: 12.2.13
  	 @angular/platform-browser-dynamic: 12.2.13
  	 @angular/router: 12.2.13
  	 @fortawesome/angular-fontawesome: 0.9.0
  	 @ionic/angular: 5.9.1
  	 @ngrx/effects: 12.5.1
  	 @ngrx/entity: 12.5.1
  	 @ngrx/router-store: 12.5.1
  	 @ngrx/schematics: 12.5.1
  	 @ngrx/store: 12.5.1
  	 @ngrx/store-devtools: 12.5.1
  	 @nxtend/capacitor: 12.0.0
  	 @nxtend/ionic-angular: 12.0.0
  	 angular-builder-custom-terser-options: 1.0.1
  	 ngx-bootstrap: 7.1.0
  	 @angular-builders/custom-webpack: 12.1.3
  	 @angular-devkit/build-angular: 12.2.13
  	 @angular/cli: 12.2.13
  	 @angular/compiler-cli: 12.2.13
  	 @angular/language-service: 12.2.13
  	 @nxtend/capacitor: 12.0.0
```

**Comment by nhhockeyplayer** [comment]
I feel like the original instantiation of the api is not getting past the original instantiation pass

something isn't doing a NEW on my model? 

not enough info is being logged to solve

**Comment by nhhockeyplayer** [comment]
this is the nestjs module

```
@Module({
    imports: [TypegooseModule.forFeature([PermissionEntity], environment.dbSchema)],
    controllers: [PermissionController],
    providers: [PermissionService]
})
export class PermissionModule {
}

```

**Comment by nhhockeyplayer** [comment]
this is the actual entity

```
@pre<PermissionEntity>('save', function(next) {
    if (!this.createdAt) {
        this.createdAt = this.updatedAt = new Date()
    }
    this.updatedAt = new Date()
    next()
})
@modelOptions({ schemaOptions: { collection: 'permission, timestamps: true' } })
export class PermissionEntity extends AbstractEntity<PermissionEntity> implements IPermissionDTO {
    @prop({ unique: true, type: () => String })
    name: string

    @prop({ name: 'display_name', type: () => String })
    displayName: string

    @prop({ type: () => String })
    description: string

    @prop({ type: RoleEntity, discriminators: () => [] })
    roles?: RoleEntity[]
}

const permissionModel = getModelForClass(PermissionEntity, { schemaOptions: { collection: 'permission' } })


```

**Comment by nhhockeyplayer** [comment]
here is my base class entity

```
@modelOptions({
    options: { allowMixed: Severity.ALLOW },
    schemaOptions: {
        timestamps: true,
        toJSON: {
            virtuals: true,
            getters: true,
        },
    },
})
@plugin(AutoIncrementSimple, [{ field: "_id" }])
export abstract class AbstractEntity<T extends { hash?: (() => {}) }> extends Integrity<T> implements ITrackable, IBusinessObject<T> {
    @prop()
    _id?: mongoose.Types.ObjectId | number | string

    @prop()
    createdBy: string

    @prop()
    createdAt: Date

    @prop()
    updatedAt: Date

    @prop()
    deletedAt: Date

    /**
     * Object Oriented Interface ctor, copy ctor, clone, equals, hash, toString
     */
    constructor(data: Partial<T> = {}) {  // copy ctor with conditional partial props
        super(data)

        Object.assign(this, data)
    }

    clone(rhs: T): T {
        return MergeService.deepMerge({}, rhs, { clone: true })
    }

    equals(rhs: T): boolean {
        return this.hash() === rhs.hash()
    }

    hash(): string {
        return this._id.toString()
    }

    toString(): string {
        return this.hash()
    }
}

```

**Comment by nhhockeyplayer** [comment]
any assist is appreciated

**Comment by smolinari** [comment]
@nhhockeyplayer -  This makes little sense to me:

```ts
@plugin(AutoIncrementSimple, [{ field: "_id" }])
export abstract class AbstractEntity<T extends { hash?: (() => {}) }> extends Integrity<T> implements ITrackable, IBusinessObject<T> {
    @prop()
    _id?: mongoose.Types.ObjectId | number | string
```

My suggestion is don't mess with the `_id` field at all. It is designed specifically for MongoDB (i.e. globally unique, which is important for horizontal scaling). So, remove the autoincrement plugin.  Also, I personally use the `id` virtual, which is standard in Mongoose. https://mongoosejs.com/docs/api.html#document_Document-id

Scott

**Comment by nhhockeyplayer** [comment]
Hi I just updated the code and boiled it down to just the role entity

**Comment by nhhockeyplayer** [comment]
ok I will try your suggestion results pending thanks friend for fast response

Im a lead consultant here in usa and nothing is cutting it for model orm and im falling back on typegoose I already have a back end server publicly released

its my OOP modeling that isnt being held up

so I hope to get this right and forge ahead with typegoose

**Comment by nhhockeyplayer** [comment]
Since your on the line here... would you happen to know of decorators for relationship that can be used with typegoose?

@OneToOne
@OneToMany
@ManyToOne
@ManyToMany

?

Thanks

**Comment by nhhockeyplayer** [comment]
well Im unsure what you want me to do with the _id property you said dont mess with it

what do I do to the code?

Is there an example?

I do not code any schema stuff in my models... do I need to ?

any example code would be a huge help and I can try your suggestion

**Comment by smolinari** [comment]
> Since your on the line here... would you happen to know of decorators for relationship that can be used with typegoose?
> 
> @OnetoOne @OneToMany @manytoone @manytomany
> 
> ?
> 
> Thanks

Mongoose or rather MongoDB doesn't have RDBMS-like relationships. You can build relationships with references, however they are only pointers to other documents in other collections. Mongoose can simulate making relationships via the `populate` method. This is all very basic Mongoose though. 

For your issue with `_id`, the discussion in this issue might help you: https://github.com/typegoose/typegoose/discussions/596

Scott

**Comment by nhhockeyplayer** [comment]
I remain stuck with the original exception happening on this ticket

my models wont hold water on bootstrap() of the api

still working it

**Comment by nhhockeyplayer** [comment]
here is the role.entity.ts

```
import { IRoleDTO } from '@workspace/shared-dto/typegoose/i-role.dto'
import { getModelForClass, modelOptions, pre, prop } from '@typegoose/typegoose'
import { AbstractEntity } from '@workspace/shared-entity/typegoose/abstract.entity'

@pre<RoleEntity>('save', function(next) {
    if (!this.createdAt) {
        this.createdAt = this.updatedAt = new Date()
    }
    this.updatedAt = new Date()
    next()
})
@modelOptions({
    schemaOptions: {
        _id: false,
        collection: 'role',
        toJSON: {
            virtuals: true,
            getters: true
        }
    }
})
export class RoleEntity extends AbstractEntity<RoleEntity> implements IRoleDTO {
    @prop({ unique: true })
    name: string

    @prop({ name: 'display_name' })
    displayName: string

    @prop({ type: () => String })
    description: string

    @prop({ name: 'is_super', type: () => Boolean })
    isSuper: boolean
}

const roleModel = getModelForClass(RoleEntity, { schemaOptions: { collection: 'role' } })

```

**Comment by nhhockeyplayer** [comment]
here is abstract.entity.ts

```
import { Integrity } from '@workspace/shared-util/integrity'
import { modelOptions, mongoose, prop, Severity } from '@typegoose/typegoose'
import { ITrackable } from '@workspace/shared-dto/typegoose/i-trackable'
import { IBusinessObject } from '@workspace/shared-dto/i-business-object'
import { MergeService } from '@service/core/merge.service'

@modelOptions({
    options: { allowMixed: Severity.ALLOW },
    schemaOptions: {
        _id: false,
        autoCreate: true,
        autoIndex: false,
        timestamps: true,
        toJSON: {
            virtuals: true,
            getters: true,
        },
    },
})
export abstract class AbstractEntity<T extends { hash?: (() => {}) }> extends Integrity<T> implements ITrackable, IBusinessObject<T> {
    @prop({
        autopopulate: true,
    })
    id?: mongoose.Types.ObjectId | number | string

    @prop()
    createdBy: string

    @prop()
    createdAt: Date

    @prop()
    updatedAt: Date

    @prop()
    deletedAt: Date

    constructor(data: Partial<T> = {}) {  // copy ctor with conditional partial props
        super(data)

        Object.assign(this, data)
    }

    clone(rhs: T): T {
        return MergeService.deepMerge({}, rhs, { clone: true })
    }

    equals(rhs: T): boolean {
        return this.hash() === rhs.hash()
    }

    hash(): string {
        return this.id.toString()
    }

    toString(): string {
        return this.hash()
    }
}

```

**Comment by nhhockeyplayer** [comment]
here is my integrity.ts

```
/**
 * Integrity Runtime Typed Class<T>
 */
import { FindOrCreate } from '@typegoose/typegoose/lib/defaultClasses'

export abstract class Integrity<T> extends FindOrCreate{
    protected constructor(private attributes: Partial<T> = {}) {
        super()

        for (const key in attributes) {
            if (attributes[key]) {
                this.attributes[key] = attributes[key]
            }
        }
    }

    static [Symbol.hasInstance](obj) {
        if (!obj) return false
        if (!!obj && !!obj.constructor) return obj.constructor.name == this.constructor.name
        else return false
    }
}
```

**Comment by nhhockeyplayer** [comment]
thats comprises the entity hierarchy

next I will post the dto interface hierarchy

**Comment by nhhockeyplayer** [comment]
here is i-role.dto.ts

```
import { IRolePermissionDTO } from './i-role-permission.dto'
import { IUserDTO } from '../typegoose/i-user.dto'
import { RoleEntity } from '@workspace/shared-entity/typegoose/role.entity'
import { IBusinessObject } from '@workspace/shared-dto/i-business-object'

export interface IRoleDTO extends IBusinessObject<RoleEntity> {
    name?: string
    displayName?: string
    description?: string
    isSuper?: boolean
}

```

**Comment by nhhockeyplayer** [comment]
here is i-business-object.ts

```
import { ITrackable } from '@workspace/shared-dto/typegoose/i-trackable'

export interface IBusinessObject<T> extends ITrackable {
    clone(rhs: T): T

    equals(rhs: T): boolean

    hash(): string

    toString(): string
}

```

**Comment by nhhockeyplayer** [comment]
here is trackable.ts

```
import { mongoose } from '@typegoose/typegoose'

/**
 * Instance Based TimeStamp Tracking
 */
export interface ITrackable {
    id?: mongoose.Types.ObjectId | number | string
    createdBy?: string;
    createdAt?: Date;
    updatedAt?: Date;
    deletedAt?: Date;
}
```

**Comment by nhhockeyplayer** [comment]
thats pretty much it, simple and to the point

any feedback would be very helpful
thanks in advance

**Comment by nhhockeyplayer** [comment]
here is the stack trace when I import the RoleModule

```
/usr/local/bin/node /Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js
Debugger listening on ws://127.0.0.1:61775/afebe200-954f-484f-8516-f060fbd6cc74
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
Waiting for the debugger to disconnect...
/Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js:725
        return this.id.toString();
                       ^

TypeError: Cannot read properties of undefined (reading 'toString')
    at Integrity.hash (/Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js:725:24)
    at Integrity.toString (/Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js:728:21)
    at new NoValidClassError (/Users/meanstack02/new/starter/node_modules/@typegoose/typegoose/lib/internal/errors.js:34:19)
    at Object.getName (/Users/meanstack02/new/starter/node_modules/@typegoose/typegoose/lib/internal/utils.js:276:39)
    at processProp (/Users/meanstack02/new/starter/node_modules/@typegoose/typegoose/lib/internal/processProp.js:18:24)
    at _buildSchema (/Users/meanstack02/new/starter/node_modules/@typegoose/typegoose/lib/internal/schema.js:38:43)
    at buildSchema (/Users/meanstack02/new/starter/node_modules/@typegoose/typegoose/lib/typegoose.js:130:41)
    at Object.getModelForClass (/Users/meanstack02/new/starter/node_modules/@typegoose/typegoose/lib/typegoose.js:73:39)
    at Object../libs/shared-entity/src/lib/typegoose/role.entity.ts (/Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js:822:31)
    at __webpack_require__ (/Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js:986:41)
    at Object../apps/api-typegoose/src/app/domain-model/role/role.controller.ts (/Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js:170:23)
    at __webpack_require__ (/Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js:986:41)
    at Object../apps/api-typegoose/src/app/domain-model/role/role.module.ts (/Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js:214:27)
    at __webpack_require__ (/Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js:986:41)
    at Object../apps/api-typegoose/src/app/app.module.ts (/Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js:50:23)
    at __webpack_require__ (/Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js:986:41)

Process finished with exit code 1

```

if I do not import RoleModule the nest api launches fine (i clustered all my 8core cpu's)

```
/usr/local/bin/node /Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js
Debugger listening on ws://127.0.0.1:61783/f11ae264-2872-4eba-8901-27b7eb6d53e6
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
Master server started on 52520
Debugger listening on ws://127.0.0.1:61786/2255e33c-5830-43ce-bdce-1084dddb6576
For help, see: https://nodejs.org/en/docs/inspector
Debugger listening on ws://127.0.0.1:61787/0f4dab7b-9a9d-45cc-9624-f97994baade8
For help, see: https://nodejs.org/en/docs/inspector
Debugger listening on ws://127.0.0.1:61790/bb1a9b38-22bc-436d-897e-64057df37754
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
Debugger listening on ws://127.0.0.1:61791/fef33a09-3ec6-4acc-afc2-abdbba161d51
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
Debugger attached.
Debugger listening on ws://127.0.0.1:61794/246869b6-2bc6-4d85-b863-318adc48221c
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
Debugger attached.
Debugger listening on ws://127.0.0.1:61796/8007f619-5efa-4d31-a21f-cf030bdd4681
For help, see: https://nodejs.org/en/docs/inspector
Debugger listening on ws://127.0.0.1:61797/82e3db48-a62c-4f16-b336-fcc927126873
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
Debugger attached.
Debugger listening on ws://127.0.0.1:61800/84e1e5cd-33c3-44c7-9be6-8475740f50fe
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
Cluster server started on 52524
Cluster server started on 52522
Cluster server started on 52525
Cluster server started on 52523
Cluster server started on 52529
Cluster server started on 52526
Cluster server started on 52527
Cluster server started on 52528
[Nest] 52525  - 12/03/2021, 6:52:31 AM     LOG [NestFactory] Starting Nest application...
[Nest] 52522  - 12/03/2021, 6:52:31 AM     LOG [NestFactory] Starting Nest application...
[Nest] 52524  - 12/03/2021, 6:52:31 AM     LOG [NestFactory] Starting Nest application...
[Nest] 52527  - 12/03/2021, 6:52:31 AM     LOG [NestFactory] Starting Nest application...
[Nest] 52526  - 12/03/2021, 6:52:31 AM     LOG [NestFactory] Starting Nest application...
[Nest] 52523  - 12/03/2021, 6:52:31 AM     LOG [NestFactory] Starting Nest application...
[Nest] 52528  - 12/03/2021, 6:52:31 AM     LOG [NestFactory] Starting Nest application...
[Nest] 52529  - 12/03/2021, 6:52:31 AM     LOG [NestFactory] Starting Nest application...
hi
hi
hi
hi
[Nest] 52525  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseModule dependencies initialized +181ms
[Nest] 52522  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseModule dependencies initialized +124ms
[Nest] 52522  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseCoreModule dependencies initialized +2ms
[Nest] 52522  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigHostModule dependencies initialized +2ms
[Nest] 52522  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] AppModule dependencies initialized +1ms
[Nest] 52522  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigModule dependencies initialized +2ms
hi
[Nest] 52524  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseModule dependencies initialized +147ms
[Nest] 52524  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseCoreModule dependencies initialized +1ms
[Nest] 52525  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseCoreModule dependencies initialized +14ms
[Nest] 52524  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigHostModule dependencies initialized +0ms
[Nest] 52525  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigHostModule dependencies initialized +1ms
[Nest] 52524  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] AppModule dependencies initialized +1ms
[Nest] 52525  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] AppModule dependencies initialized +0ms
[Nest] 52524  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigModule dependencies initialized +0ms
[Nest] 52527  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseModule dependencies initialized +129ms
[Nest] 52525  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigModule dependencies initialized +1ms
[Nest] 52527  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseCoreModule dependencies initialized +1ms
hi
[Nest] 52527  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigHostModule dependencies initialized +1ms
[Nest] 52527  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] AppModule dependencies initialized +1ms
[Nest] 52527  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigModule dependencies initialized +1ms
hi
[Nest] 52525  - 12/03/2021, 6:52:31 AM     LOG [RoutesResolver] AppController {/nest}: +14ms
[Nest] 52524  - 12/03/2021, 6:52:31 AM     LOG [RoutesResolver] AppController {/nest}: +15ms
[Nest] 52525  - 12/03/2021, 6:52:31 AM     LOG [RouterExplorer] Mapped {/nest, GET} route +2ms
hi
[Nest] 52524  - 12/03/2021, 6:52:31 AM     LOG [RouterExplorer] Mapped {/nest, GET} route +4ms
[Nest] 52522  - 12/03/2021, 6:52:31 AM     LOG [RoutesResolver] AppController {/nest}: +26ms
[Nest] 52525  - 12/03/2021, 6:52:31 AM     LOG [NestApplication] Nest application successfully started +4ms
[Nest] 52529  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseModule dependencies initialized +130ms
[Nest] 52529  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseCoreModule dependencies initialized +4ms
[Nest] 52529  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigHostModule dependencies initialized +1ms
[Nest] 52529  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] AppModule dependencies initialized +0ms
[Nest] 52529  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigModule dependencies initialized +1ms
[Nest] 52522  - 12/03/2021, 6:52:31 AM     LOG [RouterExplorer] Mapped {/nest, GET} route +9ms
[Nest] 52528  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseModule dependencies initialized +146ms
[Nest] 52525  - 12/03/2021, 6:52:31 AM     LOG Nest API Listening at http://localhost:3333/nest
[Nest] 52525  - 12/03/2021, 6:52:31 AM     LOG Mongoose TYPEGOOSE database Listening at mongodb://127.0.0.1:27017/nest
[Nest] 52523  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseModule dependencies initialized +144ms
[Nest] 52528  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseCoreModule dependencies initialized +4ms
[Nest] 52528  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigHostModule dependencies initialized +1ms
[Nest] 52523  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseCoreModule dependencies initialized +1ms
[Nest] 52528  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] AppModule dependencies initialized +0ms
[Nest] 52523  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigHostModule dependencies initialized +1ms
[Nest] 52528  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigModule dependencies initialized +1ms
[Nest] 52523  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] AppModule dependencies initialized +0ms
[Nest] 52523  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigModule dependencies initialized +1ms
[Nest] 52526  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseModule dependencies initialized +152ms
[Nest] 52524  - 12/03/2021, 6:52:31 AM     LOG [NestApplication] Nest application successfully started +17ms
[Nest] 52526  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] TypegooseCoreModule dependencies initialized +1ms
[Nest] 52526  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigHostModule dependencies initialized +0ms
[Nest] 52526  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] AppModule dependencies initialized +1ms
[Nest] 52526  - 12/03/2021, 6:52:31 AM     LOG [InstanceLoader] ConfigModule dependencies initialized +1ms
[Nest] 52522  - 12/03/2021, 6:52:31 AM     LOG [NestApplication] Nest application successfully started +15ms
[Nest] 52524  - 12/03/2021, 6:52:31 AM     LOG Nest API Listening at http://localhost:3333/nest
[Nest] 52524  - 12/03/2021, 6:52:31 AM     LOG Mongoose TYPEGOOSE database Listening at mongodb://127.0.0.1:27017/nest
[Nest] 52522  - 12/03/2021, 6:52:31 AM     LOG Nest API Listening at http://localhost:3333/nest
[Nest] 52522  - 12/03/2021, 6:52:31 AM     LOG Mongoose TYPEGOOSE database Listening at mongodb://127.0.0.1:27017/nest
[Nest] 52529  - 12/03/2021, 6:52:31 AM     LOG [RoutesResolver] AppController {/nest}: +31ms
[Nest] 52523  - 12/03/2021, 6:52:31 AM     LOG [RoutesResolver] AppController {/nest}: +25ms
[Nest] 52528  - 12/03/2021, 6:52:31 AM     LOG [RoutesResolver] AppController {/nest}: +28ms
[Nest] 52523  - 12/03/2021, 6:52:31 AM     LOG [RouterExplorer] Mapped {/nest, GET} route +3ms
[Nest] 52527  - 12/03/2021, 6:52:31 AM     LOG [RoutesResolver] AppController {/nest}: +60ms
[Nest] 52526  - 12/03/2021, 6:52:31 AM     LOG [RoutesResolver] AppController {/nest}: +27ms
[Nest] 52529  - 12/03/2021, 6:52:31 AM     LOG [RouterExplorer] Mapped {/nest, GET} route +10ms
[Nest] 52528  - 12/03/2021, 6:52:31 AM     LOG [RouterExplorer] Mapped {/nest, GET} route +8ms
[Nest] 52526  - 12/03/2021, 6:52:31 AM     LOG [RouterExplorer] Mapped {/nest, GET} route +4ms
[Nest] 52528  - 12/03/2021, 6:52:31 AM     LOG [NestApplication] Nest application successfully started +4ms
[Nest] 52529  - 12/03/2021, 6:52:31 AM     LOG [NestApplication] Nest application successfully started +5ms
[Nest] 52526  - 12/03/2021, 6:52:31 AM     LOG [NestApplication] Nest application successfully started +4ms
[Nest] 52529  - 12/03/2021, 6:52:31 AM     LOG Nest API Listening at http://localhost:3333/nest
[Nest] 52523  - 12/03/2021, 6:52:31 AM     LOG [NestApplication] Nest application successfully started +15ms
[Nest] 52528  - 12/03/2021, 6:52:31 AM     LOG Nest API Listening at http://localhost:3333/nest
[Nest] 52529  - 12/03/2021, 6:52:31 AM     LOG Mongoose TYPEGOOSE database Listening at mongodb://127.0.0.1:27017/nest
[Nest] 52528  - 12/03/2021, 6:52:31 AM     LOG Mongoose TYPEGOOSE database Listening at mongodb://127.0.0.1:27017/nest
[Nest] 52527  - 12/03/2021, 6:52:31 AM     LOG [RouterExplorer] Mapped {/nest, GET} route +15ms
[Nest] 52523  - 12/03/2021, 6:52:31 AM     LOG Nest API Listening at http://localhost:3333/nest
[Nest] 52523  - 12/03/2021, 6:52:31 AM     LOG Mongoose TYPEGOOSE database Listening at mongodb://127.0.0.1:27017/nest
[Nest] 52527  - 12/03/2021, 6:52:31 AM     LOG [NestApplication] Nest application successfully started +5ms
[Nest] 52527  - 12/03/2021, 6:52:31 AM     LOG Nest API Listening at http://localhost:3333/nest
[Nest] 52526  - 12/03/2021, 6:52:31 AM     LOG Nest API Listening at http://localhost:3333/nest
[Nest] 52526  - 12/03/2021, 6:52:31 AM     LOG Mongoose TYPEGOOSE database Listening at mongodb://127.0.0.1:27017/nest
[Nest] 52527  - 12/03/2021, 6:52:31 AM     LOG Mongoose TYPEGOOSE database Listening at mongodb://127.0.0.1:27017/nest

```

**Comment by nhhockeyplayer** [comment]
strange thing is I was a getting same thing with typeorm and thats why I ditched typeorm to go back to typegoose

but I doubt they are related they cant be but its spookey

**Comment by hasezoey** [comment]
from what i can tell, `webpack` is used, is it configured to use babel as the typescript compiler, when yes, did you apply [Known Issues: Babel](https://typegoose.github.io/typegoose/docs/guides/known-issues#babel)?

also, could you provide your tsconfig?

**Comment by hasezoey** [comment]
kinda unrelated notes:

> _id: false,

this will only have a effect when the schema is *not* the top-level schema (it only has a effect when its a subdocument)

> @prop({
>      autopopulate: true,
>  })
> id?: mongoose.Types.ObjectId | number | string

if this is the document's id, then i would recommend to not  use auto-populate on it (also, it is not a reference so auto-populate would probably not even work)

also, for this to correctly work, schema-option `id: false` should be set to disable mongoose's `.id` virtual

as for types, i would recommend to only use one of the three (in this case probably `mongoose.Types.ObjectId`)

> AbstractEntity

i generally do not recommend to use a `constructor` on typegoose classes and also methods should also not be mixed

also, from what i can tell, the error comes from `AbstractEntity.hash`, what type is the caller at that point, is it a instance of `AbstractEntity` or a instance of `mongoose.Document`?

**Comment by nhhockeyplayer** [comment]
here is the tsconfig.json

```
{
    "compileOnSave": false,
    "compilerOptions": {
        "resolveJsonModule": true,
        "allowSyntheticDefaultImports": true,
        "strict": false,
        "noImplicitAny": false,
        "typeRoots": ["node_modules/@types"],
        "types": ["jest", "node", "mocha", "chai", "chai-http"],
        "rootDir": "",
        "sourceMap": true,
        "declaration": false,
        "moduleResolution": "node",
        "emitDecoratorMetadata": true,
        "experimentalDecorators": true,
        "importHelpers": true,
        "target": "es2015",
        "module": "esnext",
        "lib": ["es2018", "dom"],
        "skipLibCheck": true,
        "skipDefaultLibCheck": true,
        "strictNullChecks": false,
        "baseUrl": ".",
        "paths": {
            "@mobile-app/app/*": ["apps/mobile-app/src/app/*"],

            "@api-typegoose/app/*": ["apps/api-typegoose/src/app/*"],
            "@api/app/*": ["apps/api/src/app/*"],
            "@core": ["libs/core/src/index.ts"],
            "@core/*": ["libs/core/src/lib/*"],
            "@energy/app/*": ["apps/energy/src/app/*"],
            "@material": ["libs/shared-material/src/index.ts"],
            "@material/*": ["libs/shared-material/src/lib/*"],
            "@mobile-tabbed/app/*": ["apps/mobile-tabbed/src/app/*"],
            "@service": ["libs/service/src/index.ts"],
            "@service/*": ["libs/service/src/lib/*"],
            "@sumitup/app/*": ["apps/sumitup/src/app/*"],
            "@workspace/api-service": ["libs/api-service/src/index.ts"],
            "@workspace/api-service/*": ["libs/api-service/src/lib/*"],
            "@workspace/env": ["libs/env/src/index.ts"],
            "@workspace/env/*": ["libs/env/src/lib/*"],
            "@workspace/isomorphic-interface": ["libs/isomorphic-interface/src/index.ts"],
            "@workspace/isomorphic-interface/*": ["libs/isomorphic-interface/src/lib/*"],
            "@workspace/shared-assets": ["libs/shared-assets/src/index.ts"],
            "@workspace/shared-assets/*": ["libs/shared-assets/src/lib/*"],
            "@workspace/shared-directives": ["libs/shared-directives/src/index.ts"],
            "@workspace/shared-directives/*": ["libs/shared-directives/src/lib/*"],
            "@workspace/shared-dto": ["libs/shared-dto/src/index.ts"],
            "@workspace/shared-dto/*": ["libs/shared-dto/src/lib/*"],
            "@workspace/shared-entity": ["libs/shared-entity/src/index.ts"],
            "@workspace/shared-entity/*": ["libs/shared-entity/src/lib/*"],
            "@workspace/shared-enums": ["libs/shared-enums/src/index.ts"],
            "@workspace/shared-enums/*": ["libs/shared-enums/src/lib/*"],
            "@workspace/shared-env": ["libs/shared-env/src/index.ts"],
            "@workspace/shared-env/*": ["libs/shared-env/src/lib/*"],
            "@workspace/shared-features": ["libs/shared-features/src/index.ts"],
            "@workspace/shared-features/*": ["libs/shared-features/src/lib/*"],
            "@workspace/shared-generics-collections": ["libs/shared-generics-collections/src/index.ts"],
            "@workspace/shared-generics-collections/*": ["libs/shared-generics-collections/src/lib/*"],
            "@workspace/shared-ionic": ["libs/shared-ionic/src/index.ts"],
            "@workspace/shared-ionic/*": ["libs/shared-ionic/src/lib/*"],
            "@workspace/shared-pipes": ["libs/shared-pipes/src/index.ts"],
            "@workspace/shared-pipes/*": ["libs/shared-pipes/src/lib/*"],
            "@workspace/shared-sass": ["libs/shared-sass/src/index.ts"],
            "@workspace/shared-sass/*": ["libs/shared-sass/src/lib/*"],
            "@workspace/shared-util": ["libs/shared-util/src/index.ts"],
            "@workspace/shared-util/*": ["libs/shared-util/src/lib/*"],
            "@workspace/shared-widgets": ["libs/shared-widgets/src/index.ts"],
            "@workspace/shared-widgets/*": ["libs/shared-widgets/src/lib/*"]
        }
    },
    "exclude": ["node_modules", "tmp"]
}

```

**Comment by nhhockeyplayer** [comment]
some feedback

removing the autopopulate didnt change anything

**Comment by nhhockeyplayer** [comment]
>also, from what i can tell, the error comes from AbstractEntity.hash, what type is the caller at that point, is it a instance of >AbstractEntity or a instance of mongoose.Document?

well Im an object oriented modeler so I keep everything at the instance to the ENTITY and if I need to operate a mongoose document its (if I need to and im not surer where to put that) in the model somewhere and delegate it internally gracefully into the instance obscured

so all I have left are just instances or collections of instances

Im trying to obscure as much gritty mongoose details as I can so I can be fluid at the object oriented modeling lane if that makes sense

and if I can achieve that then I can take off

**Comment by nhhockeyplayer** [comment]
>from what i can tell, webpack is used, is it configured to use babel as the typescript compiler, when yes, did you apply >Known Issues: Babel?

Im using angular cli wrapped by Nrwl/Nx cli which uses the typescript compiler
angular cli wraps webpack and obscures that deep away hidden appropriately

so this is pure typescript compiler with angular way by angular team which has proven best way for me

**Comment by nhhockeyplayer** [comment]
so Im just wondering if this is a class instantiation issue not necessarily an object instance issue
because this bootstrap() is not performing a NEW on any entities its just bootstrapping the model

I do have an elaborate seeding framework that will (if I uncomment that) perform a new on users, roles, permissions and seed the database with my intrinsic meta-data but thats commented out right now Im just trying to get the code to boot first

its kinda like when an angular object gets launched and the debugger has a first pass thru just initializing the code without data (passing the language compilation phase)

but somehow the hash method is being called and I see nothing in the stack trace that makes any sense of that to be happening no one is calling hash(). BUT maybe deep in the bowels of this thing someone in the compiler is performing a toString() call on the object at hand and its not holding water for that

Im going to comment out the toString method

I got 16yrs of java R&D under my belt and all these class modelings are standard, copy ctor() clone, equals, hash, toString
never an issue getting past the compiler

but since this is open source anything goes

feedback pending on tostring()

**Comment by nhhockeyplayer** [comment]
so its the Object.getName() deep within the buildSchema calls

and it must be doing a toString() on the class

still digging

**Comment by nhhockeyplayer** [comment]
it wouldnt be fair if limitations are placed on constructors() and methods anywhere

**Comment by hasezoey** [comment]
if i correctly understand, then currently something is causing to trigger a instantiation of `NoValidClassError`, which implicitly calls `toString` on the first argument (which in this case is a instance of `Integrity`), which calls `Integrity.hash` somehow and because of `this._id.toString()` it fails because `_id` is `undefined` at this point and so `toString` cannot be called

PS: i am not quite sure what the purpose of the `Integrity` call is

**Comment by nhhockeyplayer** [comment]
Integrity is just a runtime blip of code to double validate the object is what it is at runtime and no one has created anything less or more. Its useful when peeling entities off the http bridge for security and manually checking if they are what they are and have the properties they claim to suppose to have. Great for legacy integrations on portals.

Well Im just trying to get this to fit into the runtime.

I have new feedback

**Comment by nhhockeyplayer** [comment]
so for the abstract class abstract.,entity and interface i-business-object

I commented out the toString() (thats all I did) method in both the abstract class (the default impl of toString() ) and the interface method toString()
alsi commented out the clustering to avoid replicated output

I ran the server
it started barking

here is the stack trace

Im trying to decipher what it thinks now

```
/usr/local/bin/node /Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js
Debugger listening on ws://127.0.0.1:62318/9aea1dcc-4198-4c5a-ab68-ba2d90738c91
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
[Nest] 53798  - 12/03/2021, 8:33:09 AM     LOG [NestFactory] Starting Nest application...
hi
[Nest] 53798  - 12/03/2021, 8:33:09 AM     LOG [InstanceLoader] TypegooseModule dependencies initialized +64ms
[Nest] 53798  - 12/03/2021, 8:33:09 AM   ERROR [ExceptionHandler] Nest can't resolve dependencies of the RoleEntityModel (?). Please make sure that the argument energyConnection at index [0] is available in the TypegooseModule context.

Potential solutions:
- If energyConnection is a provider, is it part of the current TypegooseModule?
- If energyConnection is exported from a separate @Module, is that module imported within TypegooseModule?
  @Module({
    imports: [ /* the Module containing energyConnection */ ]
  })

Error: Nest can't resolve dependencies of the RoleEntityModel (?). Please make sure that the argument energyConnection at index [0] is available in the TypegooseModule context.

Potential solutions:
- If energyConnection is a provider, is it part of the current TypegooseModule?
- If energyConnection is exported from a separate @Module, is that module imported within TypegooseModule?
  @Module({
    imports: [ /* the Module containing energyConnection */ ]
  })

    at Injector.lookupComponentInParentModules (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/injector.js:202:19)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Injector.resolveComponentInstance (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/injector.js:157:33)
    at async resolveParam (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/injector.js:108:38)
    at async Promise.all (index 0)
    at async Injector.resolveConstructorParams (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/injector.js:123:27)
    at async Injector.loadInstance (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/injector.js:52:9)
    at async Injector.loadProvider (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/injector.js:74:9)
    at async Promise.all (index 3)
    at async InstanceLoader.createInstancesOfProviders (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/instance-loader.js:44:9)
    at async /Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/instance-loader.js:29:13
    at async Promise.all (index 7)
    at async InstanceLoader.createInstances (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/instance-loader.js:28:9)
    at async InstanceLoader.createInstancesOfDependencies (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/instance-loader.js:18:9)
    at async /Users/meanstack02/new/starter/node_modules/@nestjs/core/nest-factory.js:93:17
    at async Function.asyncRun (/Users/meanstack02/new/starter/node_modules/@nestjs/core/errors/exceptions-zone.js:22:13)
Waiting for the debugger to disconnect...

Process finished with exit code 1

```

**Comment by nhhockeyplayer** [comment]
energy is the name of my database

**Comment by nhhockeyplayer** [comment]
having a toString() method shouldnt be a showstopper though

is typegoose able to handle it?

**Comment by nhhockeyplayer** [comment]
this is what my main.ts looks like

energy is the desired mongo database
nest is the desired api root

connect string is 
`export const dbConnectString = process.env.MONGODB_URI || 'mongodb://127.0.0.1:27017/energy'
`

and to access my api
`https://localhost:3333/nest`



```
import { NestFactory } from '@nestjs/core'
import { AppModule } from './app/app.module'
import { Logger } from '@nestjs/common'
import { environment } from './environments/environment'
import { dbConnectString } from './orm/db-connect-string'

async function bootstrap() {
    const app = await NestFactory.create(AppModule, { cors: true })

    const globalPrefix = 'nest'
    app.setGlobalPrefix(globalPrefix)

    const port = process.env.PORT || 3333

    await app.listen(port, () => {
        Logger.log('Nest API Listening at http://' + environment.host + ':' + environment.port + '/' + globalPrefix)
        Logger.log('Mongoose TYPEGOOSE database Listening at ' + dbConnectString)
    })

}

bootstrap()

```

**Comment by hasezoey** [comment]
> having a toString() method shouldnt be a showstopper though
> 
> is typegoose able to handle it?

it is not really typegoose's problem that a property is accessed which is undefined, also because the error is to show that a variable that was provided is not a class, there is not real other way to get a same / value of something other than `toString`, and without that it would be hard to debug what is the problem

the easiest workaround would be `return this.id?.toString() ?? "undefined"`

---

> Error: Nest can't resolve dependencies of the RoleEntityModel (?). Please make sure that the argument energyConnection at index [0] is available in the TypegooseModule context.

as for the NestJS error, i cant help with that because i am not familiar with NestJS or typegoose-nestjs

**Comment by nhhockeyplayer** [comment]
this is my AppModule for the nest app

```
import { Module } from '@nestjs/common'

import { AppController } from './app.controller'
import { AppService } from './app.service'
import { ConfigModule } from '@nestjs/config'
import { TypegooseModule } from 'nestjs-typegoose'
import { RoleModule } from './domain-model/role/role.module'
import { dbConnectString } from '../orm/db-connect-string'

@Module({
    imports: [
        ConfigModule.forRoot({
            cache: true
        }),

        TypegooseModule.forRoot(dbConnectString),
        RoleModule
    ],
    controllers: [
        AppController
    ],
    providers: [
        AppService
    ]
})
export class AppModule {
    constructor() {
        console.log('hi')
    }
}

```

**Comment by nhhockeyplayer** [comment]
well at least we isolated to the buildSchema logic

Im adding that null check now

and continuing to shake this out

if you can leave this open until a hard explanation is found that would be great

Im going to try and rework my options with the schema as I have it modeled and see if I can do what I want to do safely

**Comment by nhhockeyplayer** [comment]
and also Im using nestjs-typegoose package which appears to still be a moving target
I dont know if anyone has had good luck with that
my other option is to go with MongooseModule instead if thats not performing setup

**Comment by smolinari** [comment]
@smolinari - I've just uploaded a new nestjs-typegoose package that is up-to-date. 

https://www.npmjs.com/package/@m8a/nestjs-typegoose

Scott

**Comment by nhhockeyplayer** [comment]
well I started to pear this down
by removing things
and leaning back on MongooseModule to see what Im missing

and it appears to fall on the building of the schema obviously in the stack trace

so when I rig this all up with MongooseModule... I see this

TS2741: Property 'schema' is missing in type 'typeof RoleModule' but required in type 'ModelDefinition'.

so My question for the typegoose folk, is there something Im missing my my entity that I need to rig like a schema property?

thanks for holding thru I really appreciate the response on this one

**Comment by nhhockeyplayer** [comment]
thanks scott I will peek at and give your upload a try today eventually

feedback pending

**Comment by hasezoey** [comment]
> so My question for the typegoose folk, is there something Im missing my my entity that I need to rig like a schema property?

just seems like not proper types are used

> and it appears to fall on the building of the schema obviously in the stack trace

yes, somehow function [`getName`](https://github.com/typegoose/typegoose/blob/c935188cde2890508d9b1db6fb6b1539b25beda3/src/internal/utils.ts#L305-L349) fails with `NoValidClassError`, this normally only happens when `getName` is improperly used or type metadata did not get correctly compiled, which can happen when using tsconfig (or commandline) option `transpileOnly`

---

also, you could activate and look through the [Debug Logs](https://typegoose.github.io/typegoose/docs/guides/advanced/logger#enable-debug-logger) from typegoose

**Comment by nhhockeyplayer** [comment]
So Im finding this appears to be open ground... modeling the schema property for this entire model ecosystem

I previously deployed a app/server angular/koa2 on typegoose 5 years back and its still running at predict34.com

I recall my models were centered on one property at the bottom of the model file

so Im thinking about adding this to my model RoleEntity

```
    static schema = getModelForClass(RoleEntity, {schemaOptions: { collection: 'role' } })
}

```

and refer to it in a static manner somewhere within Im sure I will find the spot, just to get it rigged

any thoughts?

or will the framework(s) grab it and be happy with it... the property 'schema'

still shaking this out

**Comment by nhhockeyplayer** [comment]
no luck as of yet... im going to give the new nestjs-typegoose a whirl

result pending

**Comment by nhhockeyplayer** [comment]
striking out on this...

here is latest stack trace from nestjs-typegoose

I only have one dependency RoleEntity with RoleModule isolated and to the point all details exposed above in code snippets

when things like this happen I need to review imports and search for circular dependency somewhere but Im fairly tight on those

maybe need to sit on this for day or two

if anything comes to mind please advise

I appreciate the assist

```
/usr/local/bin/node /Users/meanstack02/new/starter/dist/apps/api-typegoose/main.js
Debugger listening on ws://127.0.0.1:64090/4aa43635-c54e-4bc4-89f7-1034e80097d0
For help, see: https://nodejs.org/en/docs/inspector
Debugger attached.
[Nest] 57304  - 12/03/2021, 1:57:58 PM     LOG [NestFactory] Starting Nest application...
[Nest] 57304  - 12/03/2021, 1:57:58 PM     LOG [InjectorLogger] Nest encountered an undefined dependency. This may be due to a circular import or a missing dependency declaration.
hi
[Nest] 57304  - 12/03/2021, 1:57:58 PM   ERROR [ExceptionHandler] Nest can't resolve dependencies of the TypegooseModuleOptions (?). Please make sure that the argument dependency at index [0] is available in the TypegooseCoreModule context.

Potential solutions:
- If dependency is a provider, is it part of the current TypegooseCoreModule?
- If dependency is exported from a separate @Module, is that module imported within TypegooseCoreModule?
  @Module({
    imports: [ /* the Module containing dependency */ ]
  })

Error: Nest can't resolve dependencies of the TypegooseModuleOptions (?). Please make sure that the argument dependency at index [0] is available in the TypegooseCoreModule context.

Potential solutions:
- If dependency is a provider, is it part of the current TypegooseCoreModule?
- If dependency is exported from a separate @Module, is that module imported within TypegooseCoreModule?
  @Module({
    imports: [ /* the Module containing dependency */ ]
  })

    at Injector.resolveSingleParam (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/injector.js:141:19)
    at resolveParam (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/injector.js:108:49)
    at Array.map (<anonymous>)
    at Injector.resolveConstructorParams (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/injector.js:123:58)
    at Injector.loadInstance (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/injector.js:52:20)
    at Injector.loadProvider (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/injector.js:74:20)
    at /Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/instance-loader.js:44:62
    at Array.map (<anonymous>)
    at InstanceLoader.createInstancesOfProviders (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/instance-loader.js:44:36)
    at /Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/instance-loader.js:29:24
    at Array.map (<anonymous>)
    at InstanceLoader.createInstances (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/instance-loader.js:28:49)
    at InstanceLoader.createInstancesOfDependencies (/Users/meanstack02/new/starter/node_modules/@nestjs/core/injector/instance-loader.js:18:20)
    at /Users/meanstack02/new/starter/node_modules/@nestjs/core/nest-factory.js:93:38
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async Function.asyncRun (/Users/meanstack02/new/starter/node_modules/@nestjs/core/errors/exceptions-zone.js:22:13)
Waiting for the debugger to disconnect...

Process finished with exit code 1
```

**Comment by nhhockeyplayer** [comment]
Im really dismayed these ORM back end frameworks and libraries wont hold normal object oriented taxonomies and class hierarchies

too much following
we need pioneers that will lead and ditch the previous cave ins I see too much following in the field

OOP has been around since late 1980s

it should not be impeded like this

guess Im asking for help in this area

im trying to break ground here

any hands on deck for this is highly desirable and appreciated

the schema is not happening here

have a great weekend

**Comment by nhhockeyplayer** [comment]
just to recap
my end goal here is to have a hashable entity whereby I can call a hash function on any entity and collect them with that signature
id's are ok but for govt work they are a bit more involved and intricate

**Comment by nhhockeyplayer** [comment]
this site has a bunch of examples but they are all 5.0
https://codesandbox.io/examples/package/typegoose

im having a hard time finding real examples with latest typegoose

and getting this up and running

are there any to refer to?

**Comment by hasezoey** [comment]
> this site has a bunch of examples but they are all 5.0
> https://codesandbox.io/examples/package/typegoose

try https://codesandbox.io/examples/package/@typegoose/typegoose

Disclaimer: i never used the site `codesandbox.io` before, just applied the correct package name

**Comment by smolinari** [comment]
@nhhockeyplayer - Are you following these instructions? https://kpfromer.github.io/nestjs-typegoose/docs/usage

I'm using Typegoose with Nestjs (via the nestjs-typegoose module) with zero problems. So, keep plugging at it. 

And, you can do extension (i.e. OOP inheritance) with Typegoose/ Mongoose. However, remember you are working with a database schema and not directly OOP objects (although they come very close, which is the whole idea), so you need to conform to Mongoose's way of doing inheritance called discriminators (which is for more complex inheritance). Simple extension is also possible, for exact sharing of fields/ schema between to classes. So, for instance, you can have a base schema with fields you'd have in any number of collections and extend from that class.  

I'd also suggest moving this discussion to Nest's Mongoose channel. As you are having issues with Nest more than Typegoose. I'm there and can help.  I'd also just make a friendly suggestion, don't ramble with your thoughts in such thread/ discussions. Keep to the problem at hand. If you have a different problem, start a new thread/ discussion. 😀 

Scott

**Comment by nhhockeyplayer** [comment]
Thanks guys for all your responsiveness and being responsible to this. Its an awesome framework.

Im seeing alot of valid usage of nestjs-typegoose model loader and its found its place

But I want to cite something. If it cannot hold any instance of class or interface that really should be addressed.

So this ticket identifies a bootstrap() case where the app cannot get on its feet yet due to a method on the instance being called by coincidence and it shuts down the bootstrap()

So I wanted to see if both sides here can consider something.

If the abstract framework stands in its own abstract realm dealing with these models beit live, instance, created or just meta.... I think its fair to state at this time for this case that its a meta scenario. And I think thats fair. So that brings into the fold... reflect-metadata pkg... shouldnt that be whats trying to do what the framework(s) (typegoose or nestjs-typegoose) is attempting to do for this scenario?

Either one or the other or both could be making the call during buildSchema or maybe even mongoose itself causing the failure.

Can you guys review what either archetype (nestjs-typegoose or typegoose) can should or could do for this scenario?

certainly there could be some defensive mechanisms in place for each framework to KNOW what its dealing with and Im sure there are several use cases of what the model could be for whatever the developer is struggling to model and the appropriate action would be a detailed log message telling them 'no you cant create a model like that' instead of just a NoValidClassError
 ... it could be the NoValidClassError itself alone not using reflect-metadata like its suppose to to get the info it needs and allow the object to fold in.
 and if it is  a NoValidClassError
 then more info should be conveyed to the developer about it... are they inheriting from mongoose.Document when they sholudnt? or are they extending ssomethingthey sshouldn't
? Im thinking more smarts can be built into this with some instance checks  on what the developer is potentially trying to cat as a model.

and it ouldb e the omission of reflect-metadata itself when it sshouldbe used instead of calling a method directly on the model?


ok my keyboard is gonig south... applemacbook


just soem food for thoughrt

i will resume mor eon monday


but wanted to jar you guys to maybe consider this

thanks for all help

**Comment by nhhockeyplayer** [comment]
i will look at those docs all suggested

**Comment by nhhockeyplayer** [comment]
so this is the guy in question

Utils.ts:305
```
/**
 * Get the correct name of the class's model
 * (with suffix)
 * @param cl The Class
 * @param customOptions Extra Options that can be added in "buildSchema"
 */
export function getName<U extends AnyParamConstructor<any>>(cl: U, customOptions?: IModelOptions) {
  // this case can happen when type casting (or type being "any") happened and wanting to throw a Error (and there using "getName" to help)
  assertion(!isNullOrUndefined(cl), new NoValidClassError(cl));

  const ctor: any = getRightTarget(cl);
  const options: IModelOptions = Reflect.getMetadata(DecoratorKeys.ModelOptions, ctor) ?? {};
  const baseName: string = ctor.name;
  const customName = customOptions?.options?.customName ?? options.options?.customName;

  if (typeof customName === 'function') {
    const name = customName(options);

    assertion(
      typeof name === 'string' && name.length > 0,
      new StringLengthExpectedError(1, name, baseName, 'options.customName(function)')
    );

    return name;
  }

```
If someone could consider making typegoose not place limitations on constructors() or anything for classes as is the typescript language itself that would be appropriate.

Or maybe its doing something wrong and can do it in another fashion with reflect-metadata

all im after are no limitations placed on the typescript language and classes so object oriented modelers can do their normal automated stuff

thanks for all your graciousness

**Comment by nhhockeyplayer** [comment]
so just to recap I think its fair to say...

typegoose is calling my model/class when its not a model yet (not an instance yet) during main.bootstrap()

resulting in the undefined

if you can be iopen minded with me and side with me that this is and should be a bug because it limits what im basically trying to do is standard natural world object oriented modeling

I looked at alot of projects for nestjs-typegoose and typegoose together
they are all fine and well
I GET how models are used with it in its current state

But I am being limited here 

if we can make this the best framework on the planet it will be with no limitations on the class taxonomy/hierarchy during bootstrap()

hoping for the best
so object oriented modeling can happen with the best model lib on the planet

**Comment by nhhockeyplayer** [comment]
SO I will be developing for the next 10 days
my codebase
as it evolves Im sure it will take on new form
and I will know for sure on my end better clarity

but if you see something that can be done in the interim that would be great
I wont have my large codebase tight for another ten days though and forging it out now to get something on deck if I can

if I cant this ticket will surely convey that in ten days

so if we can leave this open for now to see what lands that would be generous

unless you see something that needs to be done to support class hierarchies in general and not limit the typescript

thanks

**Comment by smolinari** [comment]
@nhhockeyplayer - I see you ignored my point about rambling. I have absolutely NO clue what your problem is. Maybe you could just state the problem you are facing, maybe with a reproduction repo, so I can help. Otherwise, I'm out. 

Scott

**Comment by hasezoey** [comment]
like in the earlier comments, i said the error comes from when using `utils.getName`:

https://github.com/typegoose/typegoose/blob/c935188cde2890508d9b1db6fb6b1539b25beda3/src/internal/utils.ts#L305-L349

the exact line is:

https://github.com/typegoose/typegoose/blob/c935188cde2890508d9b1db6fb6b1539b25beda3/src/internal/utils.ts#L313

which you can clearly see that the input value is either `null` or `undefined`, which in turn is of course not a valid class

and `NoValidClassError` tries to output the variable via a implicit `.toString()` call (which you modified), which is entirely appropriate in my opinion, because there is not real other way to identify a value aside from very primitive `typeof`

also, from what i know `toString()` should be able to handle that not all fields of the variable are existing

---

the issue in itself is weird, because as can be seen in the earlier code mention, this error at this line should only get triggered if the input variable is `null` or `undefined`, but when trying to call `.toString()` somehow it has the overwritten function

---

if you are interested in following up on these problems, see:
- #641
- #642

**Comment by hasezoey** [comment]
typegoose 9.3.0 includes better handling of (unknown and any) variables in errors, which should progress this error somewhere else

but this should still not have been triggered, see #641

**Comment by nhhockeyplayer** [comment]
Hi Scott... my model wont boot. Thats as simple as it gets. 
Sorry if I confused you. 

I wont have better handle on my monorepo for a bit 
but I eventually will Im pre-empted right now and will come back to this to see it thru.

and I will see about a repo to show more.

Thanks for your patience

**Comment by nhhockeyplayer** [comment]
sorry guys for the latency, I been tripped up with ES Module issues
my elite framework angular has imposed this
therefore all underlying frameworks must comply
and my server nestJS is not up to par yet with ESM
so Im in a holding pattern until something gives this week

I will be back at this when my models can get booted and hopefully that will be sooner than later

**Comment by nhhockeyplayer** [comment]
while I got you folks on the line if your listening... (i will delete this clutter msg just had question)
I did a mini hibernate back in 2004 for raytheon patriot missile deployments (pdf generator crud app cookie cutter for 400 apps that drilled into 2000 oracle tables). Having taken the tech lead overnight with new architecture I modeled this using UML and delegated the backend to another consultant. The silver bullet to this was I asked the dba to write a script that, given any reference to an oracle table, that it generate and output (actual code module file) and that output being a hibernate compliant POJO. So with a click of a button I could generate the entire domain model for the back end for this SPA app. That code was retrofitted into my mini hibernate layer which generated SQL on the fly to drill into the back end. The consultant that I handed this off to used reflection (with some code that performed serialization/deserialization of any given model instance being operated on) to do this and I always wondered why he did that.

So here is my question...

I see repositories being created for typeorm and mongoose (and yes we can for typegoose if so)
and in some cases there is serialization/deserialization logic for models.
Can someone explain to me why serialization is important? Im dangerous enough to create schema and setup a postgres/mongoose/mongodb/mysql and have done model servers before.
But I tend to think this serialization is in fact essential for the layer.
Can anyone elaborate on this for me?
Because a model server I designed prior did not do that it just grabbed the model and ran with it.
What would be the benefits of serialization and downfalls if not used?
I lack the concept a bit and unsure if its even persisted to the file system or not or in DRAM.
Any thoughts help and explanation on this would be terrific.
Thanks
Ken

**Comment by smolinari** [comment]
The question about serialization would be better on the chat or here in discusssions @nhhockeyplayer.

https://discord.gg/DKjR7Fbr
https://github.com/typegoose/typegoose/discussions

Scott

**Comment by nhhockeyplayer** [comment]
Pending... got my models resolved pure OOP automated instance mapping peeled right off the http bridge
using gamma pattern
been rough week
will be shaking out typegoose soon for this likely into weekend if not sooner

**Comment by nhhockeyplayer** [comment]
The main limitation here is folks think they can operate model classes on the client side and they cant.
There are underlying mechanisms within the typegoose framework and its dependencies that rely on packages ONLY built to run on the server side

There are alot of threads from 2019 where the public was struggling with just this thinking they can operate models on the client side to no avail and there remains an open issue unresolved its a bear to attempt to get it to work

So to recap one of my issues was we can use typegoose models like this on the client side... the decorators are just not welcome
```

@pre<VideoEntity>('save', function(this, next) {
    if (!this.createdAt) {
        this.createdAt = this.updatedAt = new Date()
    }
    this.updatedAt = new Date()
    next()
})
@modelOptions({
    schemaOptions: {
        _id: false,
        collection: 'user',
        toJSON: {
            virtuals: true,
            getters: true
        }
    }
})
export class VideoEntity extends AbstractEntity<VideoEntity> implements IVideoDTO {
    @prop({ unique: true })
    id: mongoose.Types.ObjectId | number | string

    @prop({ name: 'display_name' })
    title: string
```

POJOs are not supported so thats a limitation now due to mongoose and typegoose

so in order to get them on the client side they need to be mapped to an interface or a pure pojo with no decorators that implements an interface

otherwise we get this

https://github.com/typegoose/typegoose/issues/33

cannot put typegoose models on client

in any event that recaps my model hierarchy issue

nestJS needs to repair the ES Module compliance for me to qualify further

but I will run ahead on this as far as I can into weekend with feedback for this ticket

**Comment by nhhockeyplayer** [comment]
just more feedback

If I comment out the typegoose decorators (seems to be fine with mongoose ObjectId type

```
// @modelOptions({
//     // options: { allowMixed: Severity.ALLOW },
//     schemaOptions: {
//         // foreignField: '_id',
//         // localField: 'id',
//         // _id: false,
//         autoCreate: true,
//         autoIndex: false,
//         timestamps: true,
//         toJSON: { virtuals: true, getters: true },
//         toObject: { virtuals: true, getters: true }
//     }
// })
export abstract class AbstractEntity<T extends IBusinessObject<T>> extends Integrity<T> implements IBusinessObject<T> {
    // @prop()
    id?: mongoose.Types.ObjectId | number | string

```
app runs fine

does anyone know why this is? whats causing it to fail the front end?

**Comment by smolinari** [comment]
Mongoose and indirectly Typegoose are both built for a Node environment. So, trying to shoehorn them into the browser just to use the types is going to be a futile undertaking. 

I, personally, am using GraphQL, where DTO's are the name of the game on the server. The DTO's, via decorated metadata and NestJS, are redefined automatically into GraphQL types. These, on the other hand, are regenerated as types for use in the frontend via graphql-codegenerator. So, my architecture looks like this:

![image](https://user-images.githubusercontent.com/3314957/146316197-057bfbbc-6a47-40f3-b4af-4c0901e4dabf.png)

So, I have 100% automation for the frontend types. :)

Scott

**Comment by hasezoey** [comment]
closing because this issue has gone off-topic (also the initial issue of `TypeError: Cannot read properties of undefined (reading 'toString')` has been fixed in 9.3.0 and 9.3.1)

**Comment by nhhockeyplayer** [comment]
Hi Scott, Is that part of your nest-typegoose lib? I could love to check that out.

The patch coming from this ticket seems to have fixed something. Which I asserted was it is calling my class when it isnt an instance yet. I felt something got fixed but unsure if it fixes that Im still being blocked by ESM...

The issue with ESM is if I us @angular/common in a class and then use that like a Util.ts class that automates http headers... nestjs caves in due to non-esm compliance. Angular is 100% ESM now no way out. But one thing I noticed is I stopped doing that and still receiving ESM exception and its coming from 3rd party libraries that are using @angular/common... so nestjs is like if you pick thw wrong thing and dont know if it uses @angular/common like some jwt util lib I kow there is jwt-angular it caves in.

Im not impressed with the microservices design for nestjs something isnt right with that it seems like it was just clobbed ontop without any real design thought. Thats what I cant stand about open source. and for the field to have only two choices like typeorm or typegoose is shameful in itself. the field has taken a dive since MIT X-Windows days opensource.

any kid playing z-box can create a library and the industry will jump at it just like that pivot-table 

open source needs to upgrade their registry with some standards too many human beings want to put something into the public domain to feel good take it to an extent and then reality starts to show

but it is what it is as to why I try not to follow and lead by DIY as much as possible

**Comment by smolinari** [comment]
@nhhockeyplayer 

> Is that part of your nest-typegoose lib? I could love to check that out.

No, it isn't. What is happening is, with nestjs-typegoose, you can use Typegoose with NestJS. 

NestJS, on the other hand, generates GraphQL SDL automatically from the code you write ( from either entities and/or your DTOs, your resolver classes/functions, etc.). That auto-generated GraphQL SDL is also made available for introspection i.e. via GraphQL Playground/ Graphiql/ Altair (which is what I use). And, this introspection API is what the frontend code uses (via graphql-code-generator) to create the frontend's types, queries and mutations (and only if kitchen sink queries and mutations are wanted, which they shouldn't be...otherwise they are, at a minimum, another a great reference for frontend devs above and beyond the introspection docs.)

> If you have a graphql solution that eliminates an ORM, well im unsure yet what you have brewing. Sounds interesting.

Absolutely no GraphQL solution will get rid of a datastore API layer. 

> Please advise when you have something.

I'm just putting these different tools together and as a proof-of-concept, I got the above workflow working. 

> Does it eliminate your very lib nestjs-typegoose?

I still rely on that small library (as explained above) and why I've forked it and am doing the upkeep on it now.

> Would it make sense to support orm based entities? You got me wondering.

Entities are necessary to make the mapping between the database and the data needed by the system. With Mongoose/Typegoose, or rather because of how MongoDB works, the schema defined via the entities is the schema used to run the database. This saves the step of having to keep up a schema version with your code and one of the reasons why a NoSQL database like MongoDB is so useful. Your code is your schema. You do still have some migration work here and there, but it isn't half as complicated as keeping schema with an RDBMS in sync with your code. That is also why Mongoose is called an Object Data Mapper (ODM) and not an Object Relational Mapper (ORM). 

Scott

**Comment by nhhockeyplayer** [comment]
thanks scott... the nestjs guys blocked me from creating issues... I think I ruined someones holiday. 
I didnt mean to. but alot of folks were quietly being blocked.
a review for the esm stuff is on the table now
I hope they lighten up nestJs owns the terrain and has captured the field the way it can yet but it will
I will watch for your new stuff 
If entire tiers/layers can be eliminated/transformed/automated thats awesome.
ciao

**Comment by nhhockeyplayer** [comment]
Hi Scott, Im going thru my models now and seeing which and how many points of failure Im encountering its several.

Anyway while your tuning up your pioneering I wanted to suggest maybe take a peek at a framework I use to be a committer on back in 2006/2007 founded by Chris Nelson founder of java.net he recruited me to help on that.

The framework was "trails" at the time. Its now called tynamo handled by 2-4 developers. It was the sweetest archetype on the planet at the time and only for special interest. Some of the best data structure modelings were cast within that framework. It was an end to end CRUD archetype. Hardwired to hibernate and tapestry sadly but the core feature was reflection with fundamental PROPERTY DESCRIPTORS that offered a turnkey fullstack project generator.

What did the developer get after creating a new project?

Out of the box a complete running fullstack app based on java.

But under the hood was something really neat... the PropertyDescriptor decorator and more of the same variations of such. Those custom decorators were applied static out of the box to a base set of CRUD user interface modules that come with every model/entity LIST, EDIT, VIEW, ... well I might be off but here is the main core lib

https://github.com/tynamo/tapestry-model

and with that you got what it did under the hood free UI (configurable skins included) for any model you dreamed of having in your fullstack end to end. These free UI pages came generated out of the archetype but by the several @PropertyDescriptor decorators that go heavy on reflection by analyzing each class (your actual model(s)), its properties and methods and automagic UI came free right out of the box.

What attracted me to typegoose was https://github.com/szokodiakos brilliant use of reflection and decorators

typegoose framework is already there and just needs that tiny push in the right direction. I created a ticket years back for typegoose as an enhancement to make it a platform just like tynamo/trails.

It would be nice to see this happen and awesome if it could fall directly onto IONIC

The use of reflect-metadata in typegoose is impressive but it does not go as far as what tynamo/trails did by bundling a set of CRUD/List/Edit/Query/View/Find static default pages like tynanmo/trails does and settling on a simple UI framework.

The PropertyDescriptor is the key construct in this all

If the typegoose team can seriously take a peek at this typegoose is almost there it just needs a push.

It would push typegoose as the elite framework on the planet as the constructs in tynamo/trails are the best on the planet if you take a peek at those they really are hard to resist.

They have several other libs optional to integrate like https://github.com/tynamo/tapestry-security and even a REST one also... the security lib https://github.com/tynamo/tapestry-security is pants wetting its so good... instance based security and property based security for any model.

Im not kidding these constructs would be the takeover of the entire planet of this space if integrated to typegoose and it wouldnt take much to accomplish either.

HibernateDescriptorDecorator was one part of that code core and central mechanics also

it would be nice to see this happen.

typegoose would never be caught

I highly encourage looking into this and the team seriously considering.

The property descriptor decorator and its variations/derivatives would be the way to start

**Comment by hasezoey** [comment]
> If the typegoose team can seriously take a peek at this typegoose is almost there it just needs a push.

if i understand the project `` correctly, then it is basically what typegraphql / typeorm / typegoose does, but also adds a generator for the models and also some extra default provided models (unlike the default classes typegoose already provides)

but this is out-of-scope for the (main, this repository here) typegoose project - though i would not be against if someone creates something like that and would be added to https://github.com/typegoose/

**Comment by nhhockeyplayer** [comment]
I created a discussion if anyone wants to speculate and brainstorm

https://github.com/typegoose/typegoose/discussions/652
